feat: Add service graph and eBPF drop stats to Hubble Grafana dashboard

alvidofaisal opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title, description and a `Fixes: #XXX` line if the commit addresses a particular GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then please add the commit author[s] as reviewer[s] to this issue. 
- [X] All commits are signed off. 
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md) 
- [X] Thanks for contributing!

This pull request enhances the Hubble Grafana dashboard by introducing service graph capabilities and detailed eBPF drop statistics.

Key changes include:
1.  **Modified Hubble Grafana Dashboard**: Updates the existing dashboard in `examples/kubernetes/addons/prometheus/monitoring-example.yaml`.
2.  **Service Interaction Table**: Adds a new panel to visualize service connections from Hubble flow events, effectively acting as a service graph (displaying source/destination workloads, protocol, verdict, and flow rates).
3.  **eBPF Drop Analysis Section**: Incorporates several new panels:
    *   "Aggregated eBPF Drop Counts by Reason" (`cilium_drop_count_total`).
    *   "eBPF Drop Rate by Reason & Direction" (time-series of `cilium_drop_count_total`).
    *   "Top Dropping Pods (eBPF)" (`cilium_drop_count_total`).
    *   "Hubble Policy Denials - Source & Destination" (`hubble_drop_total`).
4.  **Dashboard Title Update**: Renamed to "Hubble Service and Drop Analysis".
5.  **Utility Script and Tests**: Includes the `update_yaml_configmap.py` script to automate dashboard updates within the ConfigMap and corresponding unit tests (`test_update_yaml_configmap.py`).

These additions aim to provide users with deeper insights into service connectivity and eBPF-level packet drop events directly within Grafana.

Fixes: 

```release-note
feat: Enhance Hubble Grafana dashboard with service graph and eBPF drop statistics

This update augments the Hubble Grafana dashboard with a service interaction table (service graph) and a comprehensive eBPF drop analysis section. Users can now visualize service connectivity and gain insights into packet drops at the eBPF level, including reasons, rates, and top affected pods. The `update_yaml_configmap.py` script is also included to facilitate easier updates to the dashboard ConfigMap.
```

tommyp1ckles commented on COMMIT_MSG.tmp:
```diff
@@ -0,0 +1,14 @@
+feat: Add service graph and eBPF drop stats to Hubble Grafana dashboard
```
was this meant to be ignored from the commit?

(also same question for several other files, such as the python scripts)

alvidofaisal replied:
Cleaned up all the unwanted files and organized everything properly in the tools directory. Thanks for the feedback.

alvidofaisal committed:
monitoring: enhance Hubble Grafana dashboard with service graph and eBPF drop stats

This update adds:
- Service interaction table showing service connections from Hubble flow events
- eBPF drop analysis section with drop counts, rates, and top affected pods
- Hubble policy denial visualization
- Dashboard title updated to 'Hubble Service and Drop Analysis'

Signed-off-by: alvidofaisal <alvidofaisal@gmail.com>

alvidofaisal committed:
examples: Add dashboard management tools for Grafana monitoring

Add comprehensive tooling for managing Grafana dashboards within the
Prometheus monitoring example. This addresses maintainer requests to
properly organize dashboard management utilities.

Key changes:
- Create tools/ directory under examples/kubernetes/addons/prometheus/
- Add Python scripts for extracting and updating dashboard configurations
- Include proper Makefile integration for dashboard operations
- Add comprehensive documentation and unit tests
- Follow Cilium coding standards with copyright headers

The tools enable developers to:
- Extract dashboard JSON from monitoring-example.yaml ConfigMaps
- Update Hubble dashboard with enhanced service graph capabilities
- Validate and test dashboard configurations
- Manage dashboard development workflow efficiently

All scripts include proper error handling, comprehensive tests, and
follow Python best practices. The Makefile provides convenient
commands accessible from the main prometheus directory.

Signed-off-by: User <user@example.com>
Signed-off-by: alvidofaisal <alvidofaisal@gmail.com>

alvidofaisal committed:
Merge branch 'main' into hubble-grafana-service-graph-ebpf-drops

Signed-off-by: alvidofaisal <alvidofaisal@gmail.com>

alvidofaisal committed:
Merge branch 'main' into hubble-grafana-service-graph-ebpf-drops

Signed-off-by: alvidofaisal <alvidofaisal@gmail.com>

tommyp1ckles commented on examples/kubernetes/addons/prometheus/README.md:
````diff
@@ -59,3 +59,41 @@ kubectl -n cilium-monitoring port-forward service/prometheus --address 0.0.0.0 -
+
+```bash
+# Show all available dashboard management commands
+make dashboard-tools
````
This expects python dependencies, should those be in a file or in a virtualenv?

tommyp1ckles commented on examples/kubernetes/addons/prometheus/tools/Makefile:
```diff
@@ -0,0 +1,112 @@
+	@chmod +x *.py
+	@echo "Tools are ready to use!"
+
+# Development workflow: extract -> edit -> update
```
One thing I'm wondering about is how we validate commits such that the checked-in monitoring-example is in sync with the extracted json files.  We should probably run the update command and check if the branch is dirty in CI to check this.

As an example, we do this with the cilium docs by run the doc generate command in CI and checking if there are any changes from HEAD.

tommyp1ckles commented on examples/kubernetes/addons/prometheus/tools/extract_dashboards.py:
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
```
These look great and I'm very much on board with adding this development workflow, however I am a little bit concerned regarding the use of Python for these scripts, specifically for ongoing maintenance of the scripts.

Zooming out and looking at what the goal we're trying to achieve here is; is provide a development workflow such that we can easily extract, update and "recompile" our monitoring-example.yaml file.  It feels like this problem may already be well solved by helm, for example:

* `helm/`
  * `templates`
    * `prometheus-k8s.yaml`
    * `grafana-config.yaml`
  * `dashboards/`
    * `cilium_dashboard.json`
    * etc...
 
Then we could generate the final "monitoring-example.yaml" using helms templating functionality.

wdyt - Is there anything here we can't do with helm?
